### PR TITLE
Improvement: improvement 3dp multi model one time import

### PR DESCRIPTION
**Pull request by MaShizhen**

### CODE REVIEW

File: src/app/flux/printing/index.js [hunk 1]
```diff
@@ -2393,143 +2408,185 @@ export const actions = {
...
-                    }
-                };
-                createLoadModelWorker(uploadPath, onMessage);
+                    };
+                    createLoadModelWorker(uploadPath, onMessage);
+                }
+            });
+        });
+
+        await Promise.allSettled(ps);
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:2567]: what if `Promise` failed

File: src/app/lib/socket-controller.js [hunk 2]
```diff
@@ -24,8 +24,7 @@ class SocketController {
         this.socket && this.socket.destroy();
 
         this.socket = io.connect('', {
-            query: `token=${token}`,
-            transports: ['websocket']
```
Comment (xiekun1992) [inline src/app/lib/socket-controller.js:28]: why delete this line

File: src/app/ui/layouts/AppLayout.jsx [hunk 3]
```diff
@@ -745,11 +746,16 @@ class AppLayout extends PureComponent {
...
-                    const file = await UniApi.Dialog.showOpenFileDialog(type);
-                    if (!file) {
+                    const isMultiSelect = pathname === '/printing';
+                    const files = await UniApi.Dialog.showOpenFileDialog(type, isMultiSelect);
+                    if (!files) {
                         return;
                     }
-                    fileObj = UniApi.File.constructFileObj(file.path, file.name.split('\\').pop());
+                    if (isMultiSelect) {
+                        fileObj = files.map(file => UniApi.File.constructFileObj(file.path, file.name.split('\\').pop()));
```
Comment (xiekun1992) [inline src/app/ui/layouts/AppLayout.jsx:755]: `file.name.split('\\')` will this works on Linux and MacOS?

File: src/app/flux/printing/actions-ui.jsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import React from 'react';
+import modal from '../../lib/modal';
+import i18n from '../../lib/i18n';
+import { Button } from '../../ui/components/Buttons';
+
+export const uiActions = {
```
Comment (xiekun1992) [inline src/app/flux/printing/actions-ui.jsx:6]: we should not access UI widgets from `flux`, place these widges into UI layer and use variables to control their visibility

**Review by xiekun1992 — CHANGES_REQUESTED** [review]

File: src/app/flux/printing/index.js [hunk 5]
```diff
@@ -2274,12 +2279,19 @@ export const actions = {
...
+    scaleToFitSelectedModel: (models) => (dispatch, getState) => {
         const { modelGroup } = getState().printing;
         const { size } = getState().machine;
         dispatch(actions.recordModelBeforeTransform(modelGroup));
 
-        const modelState = modelGroup.scaleToFitSelectedModel(size);
+        const modelState = (() => {
+            if (models && models.length > 0) {
+                const res = modelGroup.scaleToFitFromModel(size, models);
+                return res;
```
Comment (jane-rose) [inline src/app/flux/printing/index.js:2290]: extra param

File: src/app/flux/printing/index.js [hunk 6]
```diff
@@ -2393,143 +2405,190 @@ export const actions = {
...
+            const isLarge = ['x', 'y', 'z'].some((key) => modelSize[key] >= size[key]);
+
+            if (isLarge) {
+                promptTasks.push({
+                    status: 'needScaletoFit',
+                    model
+                });
             }
         });
+        if (!(modelNames.length === 1 && newModels.length === 0)) {
```
Comment (jane-rose) [inline src/app/flux/printing/index.js:2585]: change this

**Review by jane-rose — APPROVED** [review]

**Pull request merged**